lib/util.c

```diff
@@ -2193,3 +2193,137 @@ EXPORTED char *modseqtoa(modseq_t modseq)
     buf_printf(&buf, MODSEQ_FMT, modseq);
     return buf_release(&buf);
 }
+
+static char *_xsyslog_ev_escape(const char *val)
+{
+    struct buf buf = BUF_INITIALIZER;
+    int needs_escaping = 0, orig_len, escaped_len;
+    const char *p;
+
+    buf_setcstr(&buf, val);
+
+    escaped_len = orig_len = buf_len(&buf);
+
+    // Make sure the empty string is visible
+    if (orig_len == 0) {
+        buf_setcstr(&buf, "\"\"");
+        return buf_release(&buf);
+    }
+
+    for (p = buf_cstring(&buf); *p; p++) {
+        switch (*p) {
+        case '\\':
+        case '\"':
+        case '\n':
+        case '\r':
+            ++escaped_len;  // add 1 for the backslash
+
+        // FALL THROUGH
+        case ' ':
+            needs_escaping = 1;
+            break;
+        }
+    }
+
+    if (needs_escaping) {
+        char *q;
+
+        escaped_len += 2;  // add 2 for surrounding DQUOTEs
+
+        buf_truncate(&buf, escaped_len);  // grow the buffer to escaped length
```

Comment:
while buf_truncate does what it's supposed to do here, I would say that calling `buf_ensure` would make intention clearer (probably then even without the comment).

Reply:
I think truncate/ensure have unfortunate names. But this looks clearer (to me) than:
```
buf_ensure(&buf, escaped_len - orig_len);
```

Rather than do the math, using the precomputed value seems most sensible?

Reply:
Yes, let's keep it as-is, I forgot about buf_ensure having this weird API.